sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/HashAggregateExec.scala

```diff
@@ -128,6 +128,16 @@ case class HashAggregateExec(
   // all the mode of aggregate expressions
   private val modes = aggregateExpressions.map(_.mode).distinct
 
+  // This is for testing final aggregate with number-of-rows-based fall back as specified in
+  // `testFallbackStartsAt`. In this scenario, there might be same keys exist in both fast and
+  // regular hash map. So the aggregation buffers from both maps need to be merged together
+  // to avoid correctness issue.
+  //
+  // This scenario only happens in unit test with number-of-rows-based fall back.
+  // There should not be same keys in both maps with size-based fall back in production.
+  private val isTestFinalAggregateWithFallback: Boolean = testFallbackStartsAt.isDefined &&
```

Comment:
instead of merging the hash maps, shall we fix the number-of-rows-based fallback to make it similar to the size-based fallback?

Reply:
@cloud-fan - I was thinking the same way too. I found it's quite hard to fix the fallback logic. I tried the approach to add a find(key): Boolean method in generated first level map, and to first check if key already exists in first level map. But I found other case like the key can be put into second level map, later added to first level map as well (fallback row counter reset to 0 case).

Reply:
I haven't touched this part of the code for a while, can you briefly introduce how size-based fallback work?

Reply:
@cloud-fan - sure.

With an internal config `spark.sql.TungstenAggregate.testFallbackStartsAt`, we can set (1). when to fallback from first level hash map to second level hash map, and (2). when to fallback from second level hash map to sort.

Suppose `spark.sql.TungstenAggregate.testFallbackStartsAt` = "2, 3".

Then the generated code per input row (aggregate the row into hash map) looks like:

```
UnsafeRow agg_buffer = null;

if (counter < 2) {
  // 1st level hash map
  agg_buffer = fastHashMap.findOrInsert(key);
}
if (agg_buffer == null) {
  // generated. code for key in unsafe row format
  ...
  if (counter < 3) {
    // 2nd level hash map
    agg_buffer = regularHashMap.getAggregationBufferFromUnsafeRow(key_in_unsafe_row, ...);
  }
  if (agg_buffer == null) {
    // sort-based fallback
    regularHashMap.destructAndCreateExternalSorter();
    ...
    counter = 0;
  }
}
counter += 1;
```

Example generated code is Line 187-232 in https://gist.github.com/c21/d0f704c0a33c24ec05387ff4df438bff .

I tried to add a method `fastHashMap.find(key): boolean`, and change code like this:

```
...
if (fastHashMap.find(key) || counter < 2) {
  // 1st level hash map
  agg_buffer = fastHashMap.findOrInsert(key);
}
...
```

But I later found the case as I mentioned above:
1. key(a) is inserted into second level hash map (when counter exceeds 1st threshold)
2. sort-based fallback happens, and counter is reset to 0 (when counter exceeds 2nd threshold)
3. key(a) is not in first level hash map, and counter does not exceed 1st threshold, the key(a) is inserted into first level hash map as well by mistake.

We can further add code like this:

```
if ((fastHashMap.find(key) && !regularHashMap.find(key_in_unsafe_row)) || counter < 2) {
  // 1st level hash map
  agg_buffer = fastHashMap.findOrInsert(key);
}
```

But it introduces more ad-hoc change and looks pretty ugly with a lot of code needs to be moved.

Reply:
@cloud-fan - sorry, I overlooked your question, you are asking how size-based fallback works.

Size-based fallback works as:
1. try to insert into 1st level hash map, and fallback to 2nd level hash map when no space in the required memory page (`RowBasedKeyValueBatch `) - https://github.com/apache/spark/blob/master/sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/RowBasedHashMapGenerator.scala#L165-L166 .
2. try to insert into 2nd level hash map, and fallback to sort-based when no space in `UnsafeFixedWidthAggregationMap` - https://github.com/apache/spark/blob/master/sql/core/src/main/java/org/apache/spark/sql/execution/UnsafeFixedWidthAggregationMap.java#L148-L150 .
3. the 2nd level hash map will be sorted and spilled and another new 2nd level hash map will be created. The 1st level hash map cannot be spilled.

Reply:
It seems to me that the major issue is we use a single counter to control both the fast and regular hash map fallback. My first thought is to add a dedicated counter for the fast hash map fallback, then I realized that the fast hash map has a capacity property. Can we simply set the capacity to `testFallbackStartsAt._1`?

Reply:
@cloud-fan - maybe I am missing something but not sure how these two solutions fix the problem.

1. dedicated counters for two maps
```
if (counter1 < 2) {
  // 1st level hash map
  agg_buffer = fastHashMap.findOrInsert(key);
}
if (agg_buffer == null) {
  // generated. code for key in unsafe row format
  ...
  if (counter2 < 3) {
    // 2nd level hash map
    agg_buffer = regularHashMap.getAggregationBufferFromUnsafeRow(key_in_unsafe_row, ...);
  }
  if (agg_buffer == null) {
    // sort-based fallback
    regularHashMap.destructAndCreateExternalSorter();
    ...
    counter2 = 0;
  }
}
counter1 += 1;
counter2 += 1;
```

Counter example:

```
1. key_a is inserted into 1st level map (counter1 = 0)
2. a couple of keys are inserted into 1st level map (count1 =2)
3. key_a is inserted into 2nd level map (count1 = 2, count2 = 2)
```

2. set 1st level map `bitMaxCapacity` to be log2(`testFallbackStartsAt._1`).
```
if (counter < 2) {
  // 1st level hash map
  agg_buffer = fastHashMap.findOrInsert(key);
}
if (agg_buffer == null) {
  // generated. code for key in unsafe row format
  ...
  if (counter < 3) {
    // 2nd level hash map
    agg_buffer = regularHashMap.getAggregationBufferFromUnsafeRow(key_in_unsafe_row, ...);
  }
  if (agg_buffer == null) {
    // sort-based fallback
    regularHashMap.destructAndCreateExternalSorter();
    ...
    counter = 0;
  }
}
counter += 1;
```

Counter example:

```
1. key_a is inserted into 1st level map (counter = 0)
2. a couple of NULL keys are inserted into 2nd level map (count = 2). Note: 1st level map does not support NULL key.
3. key_a is inserted into 2nd level map (count1 = 2)
```

Reply:
My idea is to simulate the size-based fallback: "no space" -> "reach the capacity/limit"

Reply:
@cloud-fan - updated per offline discussion. Changed the first level fallback by restricting first level map capacity.

Reply:
Also just FYI - I updated the generated code in PR description for checking if needed.